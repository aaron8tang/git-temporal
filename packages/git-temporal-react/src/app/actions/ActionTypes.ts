export enum ActionTypes {
  REQUEST_COMMITS = 'REQUEST_COMMITS',
  RECEIVE_COMMITS = 'RECEIVE_COMMITS',
  SELECT_PATH = 'SELECT_PATH',
  INVALIDATE_PATH = 'INVALIDATE_PATH',
  HIGHLIGHT_COMMIT = 'HIGHLIGHT_COMMIT',
  VIEW_COMMITS = 'VIEW_COMMITS',
  VIEW_FILES = 'VIEW_FILES',
  ADD_AUTHOR_FILTER = 'ADD_AUTHOR_FILTER',
  REMOVE_AUTHOR_FILTER = 'REMOVE_AUTHOR_FILTER',
  REMOVE_ALL_AUTHOR_FILTERS = 'REMOVE_ALL_AUTHOR_FILTERS',
}
