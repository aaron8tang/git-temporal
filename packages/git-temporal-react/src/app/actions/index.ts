import { ActionTypes } from './ActionTypes';

export const selectPath = path => (dispatch, _getState) => {
  // if this comes from a rename, follow the most current name
  const actualPath = path.replace(/\{(.*)\s=>\s(.*)\}/g, '$2');

  dispatch(fetchCommitsIfNeeded(actualPath));
  return { selectedPath: actualPath, type: ActionTypes.SELECT_PATH };
};

export const invalidatePath = path => ({
  selectedPath: path,
  type: ActionTypes.INVALIDATE_PATH,
});

export const addAuthorFilter = authorName => ({
  authorName,
  type: ActionTypes.ADD_AUTHOR_FILTER,
});

export const removeAuthorFilter = authorName => ({
  authorName,
  type: ActionTypes.REMOVE_AUTHOR_FILTER,
});

export const removeAllAuthorFilters = () => ({
  type: ActionTypes.REMOVE_ALL_AUTHOR_FILTERS,
});

export const highlightCommit = commitId => ({
  commitId,
  type: ActionTypes.HIGHLIGHT_COMMIT,
});

export const viewCommits = () => ({
  type: ActionTypes.VIEW_COMMITS,
});

export const viewFiles = () => ({
  type: ActionTypes.VIEW_FILES,
});

export const requestCommits = path => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_COMMITS,
});

export const receiveCommits = (path, json) => ({
  selectedPath: path,
  commits: json.commits,
  type: ActionTypes.RECEIVE_COMMITS,
});

const fetchCommits = path => dispatch => {
  dispatch(requestCommits(path));
  const pathParam = path && path.trim().length > 0 ? `?path=${path}` : '';
  // TODO : replace this with serviceBaseUrl when it is in
  return fetch(`http://localhost:11966/git-temporal/history${pathParam}`)
    .then(response => response.json())
    .then(json => dispatch(receiveCommits(path, json)));
};

const shouldFetchCommits = (state, path) => {
  if (state.isFetching) {
    return false;
  }
  return (
    state.didInvalidate ||
    !state.commits ||
    state.commits.length <= 0 ||
    state.selectedPath !== path
  );
};

export const fetchCommitsIfNeeded = path => (dispatch, getState) => {
  if (shouldFetchCommits(getState(), path)) {
    return dispatch(fetchCommits(path));
  }
};
