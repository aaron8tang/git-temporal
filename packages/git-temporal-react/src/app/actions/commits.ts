import { ActionTypes } from 'app/actions/ActionTypes';
import { isVscode, vscode } from 'app/actions/vscode';
import { debug } from '@git-temporal/logger';

export const requestCommits = path => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_COMMITS,
});

export const receiveCommits = (path, json) => ({
  selectedPath: path,
  commits: json.commits,
  isFileSelected: json.isFile,
  type: ActionTypes.RECEIVE_COMMITS,
});

const fetchCommits = path => dispatch => {
  dispatch(requestCommits(path));
  if (isVscode) {
    debug('sending history request to vscode ', path);
    // see actions/vscode.ts for response handling that comes as a window event
    vscode.postMessage({ path, command: 'history' });
  } else {
    const pathParam = path && path.trim().length > 0 ? `?path=${path}` : '';
    // TODO : replace this with serviceBaseUrl when it is in
    fetch(`http://localhost:11966/git-temporal/history${pathParam}`)
      .then(response => response.json())
      .then(json => dispatch(receiveCommits(path, json)));
  }
};

const shouldFetchCommits = (state, path) => {
  if (state.isFetching) {
    return false;
  }
  return (
    !state.commits || state.commits.length <= 0 || state.selectedPath !== path
  );
};

export const fetchCommitsIfNeeded = path => (dispatch, getState) => {
  if (shouldFetchCommits(getState(), path)) {
    return dispatch(fetchCommits(path));
  }
};
