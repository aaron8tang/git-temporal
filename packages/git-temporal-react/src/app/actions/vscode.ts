import { receiveCommits, receiveCommitRange } from 'app/actions/commits';
import { receiveDiff } from 'app/actions/diff';
import { debug } from 'app/utilities/logger';

// @ts-ignore
export const isVscode = window && window.IS_VSCODE_WEBVIEW;
// @ts-ignore
export const vscode = isVscode ? acquireVsCodeApi() : null;

if (isVscode) {
  debug('running in VSCode.');
}

export function handleVscodeMessages(dispatch) {
  // @ts-ignore
  if (!window || !window.IS_VSCODE_WEBVIEW) {
    return;
  }
  window.addEventListener('message', event => {
    const { data } = event;
    debug(`actions/vscode received window message`, {
      ...data,
      // don't dump full commits or file contents to log
      commits: data.commits ? `object[${data.commits.length}]` : undefined,
      leftFileContents: redactArray(data.leftFileContents),
      rightFileContents: redactArray(data.rightFileContents),
      modifiedFiles: redactArray(data.modifiedFiles),
    });

    switch (data.type) {
      case 'commitRange':
        dispatch(receiveCommitRange(data.path, data));
        break;
      case 'history':
        dispatch(receiveCommits(data.path, data));
        break;
      case 'diff':
        dispatch(receiveDiff(data.path, data));
        break;
    }
  });
}

function redactArray(array) {
  return array ? `[..] length=${array.length}` : undefined;
}
