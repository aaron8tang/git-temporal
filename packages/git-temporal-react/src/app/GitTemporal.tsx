import React, { Component } from 'react';
import { connect } from 'react-redux';
import { debug } from '@git-temporal/logger';

import { GitTemporalProps, DispatchProps, StateProps } from 'app/interfaces';
import { selectPath } from 'app/actions';
import { handleVscodeMessages } from 'app/actions/vscode';
import { getGitTemporalContainerState } from 'app/selectors';
import { style } from 'app/styles';

import Header from 'app/containers/Header';
import Stats from 'app/containers/Stats';
import Authors from 'app/containers/Authors';
import Files from 'app/containers/Files';
import Commits from 'app/containers/Commits';
import Timeplot from 'app/containers/Timeplot';
import DifferenceViewer from 'app/containers/DifferenceViewer';

import { SpinnerContainer } from 'app/components/SpinnerContainer';
import { HorizontalScroller } from 'app/components/HorizontalScroller';
import { TransitionVisible } from 'app/components/TransitionVisible';

const outerScrollStyle = {
  _extends: 'flexRows',
  position: 'relative',
  minHeight: '100%',
  height: 'initial',
};

const innerScrollStyle = {
  position: 'relative',
  display: 'flex',
  flexDirection: 'row',
  flexGrow: 1,
};

const scrollerIconStyle = {
  transform: 'scaleY(8)',
  top: '35%',
};

const transitionStyle = {
  _extends: 'flexColumns',
  flexGrow: 1,
};

export class GitTemporal extends Component<
  GitTemporalProps & DispatchProps & StateProps
> {
  componentDidMount() {
    const { path, dispatch } = this.props;
    dispatch(selectPath(path));
    handleVscodeMessages(dispatch);
  }

  componentWillUnmount() {
    debug('unmounting GitTemporal');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.path !== this.props.path) {
      const { dispatch, path } = nextProps;
      dispatch(selectPath(path));
    }
  }

  render() {
    const { isFetching, viewCommitsOrFiles } = this.props;
    debug('rendering GitTemporal');

    return (
      <div style={style('page')}>
        <SpinnerContainer isSpinning={isFetching}>
          <div style={style('flexColumns', { height: '100%' })}>
            <Header />
            <TransitionVisible
              isVisible={!isFetching}
              style={style(transitionStyle)}
            >
              <Stats />
              <div style={{ display: 'flex', flexGrow: 1 }}>
                <HorizontalScroller
                  style={style(outerScrollStyle)}
                  innerStyle={style(innerScrollStyle)}
                  iconStyle={style(scrollerIconStyle)}
                >
                  <Authors />
                  {viewCommitsOrFiles === 'files' ? <Files /> : <Commits />}
                  <DifferenceViewer />
                </HorizontalScroller>
              </div>
              <Timeplot />
            </TransitionVisible>
          </div>
        </SpinnerContainer>
      </div>
    );
  }
}

export default connect(getGitTemporalContainerState)(GitTemporal);
