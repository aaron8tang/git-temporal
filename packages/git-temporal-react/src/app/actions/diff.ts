import { ICommit } from 'app/interfaces';
import { dateFilteredCommits } from 'app/selectors/dates';
import { ActionTypes } from 'app/actions/ActionTypes';
import { isVscode, vscode } from 'app/actions/vscode';
import { debug } from '@git-temporal/logger';
import { start } from 'repl';

const requestDiff = (path: string) => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_DIFF,
});

export const receiveDiff = (path: string, diff: any) => ({
  diff,
  selectedPath: path,
  type: ActionTypes.RECEIVE_DIFF,
});

export const setDiffStartCommit = (commitId: string) => ({
  commitId,
  type: ActionTypes.SET_DIFF_START_COMMIT,
});

export const setDiffEndCommit = (commitId: string) => ({
  commitId,
  type: ActionTypes.SET_DIFF_END_COMMIT,
});

export const fetchDiff = (
  path: string,
  commits: ICommit[],
  startDate: number,
  endDate: number
) => (dispatch: any): void => {
  dispatch(requestDiff(path));
  const filteredCommits = dateFilteredCommits(commits, startDate, endDate);
  if (!filteredCommits || filteredCommits.length === 0) {
    return;
  }
  // commits are in descending order
  const rightCommit = filteredCommits[0];
  const leftCommit = filteredCommits.slice(-1)[0];

  if (isVscode) {
    debug('sending diff request to vscode ', path, leftCommit, rightCommit);
    // see actions/vscode.ts for response handling that comes as a window event
    vscode.postMessage({
      path,
      leftCommit: leftCommit.id,
      rightCommit: rightCommit.id,
      command: 'diff',
    });
  } else {
    const pathParam =
      path && path.trim().length > 0 ? `?path=${path}` : '?path=.';
    const leftCommitParam = leftCommit ? `&leftCommit=${leftCommit.id}` : '';
    const rightCommitParam = rightCommit
      ? `&rightCommit=${rightCommit.id}`
      : '';

    // TODO : replace this with serviceBaseUrl when it is in
    const url = `http://localhost:11966/git-temporal/diff${pathParam}${leftCommitParam}${rightCommitParam}`;
    fetch(url)
      .then(response => response.json())
      .then(diff => dispatch(receiveDiff(path, diff)));
  }
};
