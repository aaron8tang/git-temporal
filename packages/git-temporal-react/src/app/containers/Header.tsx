import React, { Component } from 'react';
import { connect } from 'react-redux';

import { DispatchProps, IHeaderContainerState } from 'app/interfaces';
import { getHeaderContainerState } from 'app/selectors';
import { style } from 'app/styles';
import { selectPath, setStartDate, setEndDate } from 'app/actions';
import { EpochDateTime } from 'app/components/EpochDateTime';
import { ExplodeOnChange } from 'app/components/ExplodeOnChange';

import Search from 'app/containers/Search';

const outerStyle = {
  _extends: ['panel', 'flexColumns'],
  flexShrink: 0,
};

const appNameStyle = {
  _extends: ['inlineBlock', 'h1Text'],
  marginBottom: 10,
};

const topRowStyle = {
  _extends: 'flexRows',
  maxWidth: 1100,
};

const statsAndSearchStyle = {
  _extends: ['inlineBlock', 'flexColumns'],
  marginBottom: 10,
  flexGrow: 1,
};

const dateStyle = {
  transition: 'all 2s ease -in -out',
};

const dateSelectedStyle = {
  _extends: 'h2Text',
  margin: '0px 5px',
  color: '@colors.selected',
};

const dateOptions = {
  month: 'long',
  timeZoneName: 'short',
};

export class Header extends Component<IHeaderContainerState & DispatchProps> {
  render() {
    const { startDate, endDate, isDefaultDates } = this.props;
    const epochStyle = [dateStyle, isDefaultDates ? {} : dateSelectedStyle];
    return (
      <div style={style(outerStyle)}>
        <div style={style(topRowStyle)}>
          <div style={style(appNameStyle)}>Git Temporal </div>
          <div style={style(statsAndSearchStyle)}>
            <div>
              <Search />
              <div style={style('h2Text', { marginBottom: 10 })}>
                Stats for {this.renderPathLinks()}
              </div>
            </div>
          </div>
        </div>
        <div style={style('h5Text')}>
          From{' '}
          <ExplodeOnChange value={startDate}>
            <EpochDateTime
              value={startDate}
              displayOptions={dateOptions}
              style={style(epochStyle)}
            />
          </ExplodeOnChange>{' '}
          to{' '}
          <ExplodeOnChange value={endDate}>
            <EpochDateTime
              value={endDate}
              displayOptions={dateOptions}
              style={style(epochStyle)}
            />
          </ExplodeOnChange>
          {!isDefaultDates && (
            <span>
              {'  '}(
              <span style={style('link')} onClick={this.onResetDatesClick}>
                reset
              </span>
              )
            </span>
          )}
        </div>
      </div>
    );
  }
  renderLinkPart(part, index, fullPath, lastIndex) {
    const styles: any = [
      {
        margin: '0px 2px',
      },
    ];
    let onClick = undefined;
    if (index !== lastIndex) {
      styles.push('link');
      onClick = () => this.onLinkPartClick(fullPath);
    }
    const sep = index === 0 ? '' : '/';

    return (
      <span style={{ wordBreak: 'break-all' }}>
        {sep}
        <span style={style(styles)} key={index} onClick={onClick}>
          {part}
        </span>
      </span>
    );
  }
  renderPathLinks() {
    const { selectedPath } = this.props;
    let parts = ['repository:/'];
    if (selectedPath && selectedPath.trim().length > 0) {
      parts = parts.concat(selectedPath.split('/'));
    }
    const lastIndex = parts.length - 1;
    let fullPathSoFar = '';
    return parts.map((part, index) => {
      // > 1 means don't add 'repository:'
      if (index > 0) {
        const sep = index === 1 ? '' : '/';
        fullPathSoFar += `${sep}${part}`;
      }
      return this.renderLinkPart(part, index, fullPathSoFar, lastIndex);
    });
  }

  onLinkPartClick = fullPath => {
    this.props.dispatch(selectPath(fullPath));
  };

  onResetDatesClick = () => {
    this.props.dispatch(setStartDate(null));
    this.props.dispatch(setEndDate(null));
  };
}

export default connect(getHeaderContainerState)(Header);
