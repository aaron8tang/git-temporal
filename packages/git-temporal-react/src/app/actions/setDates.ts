import { debug } from '@git-temporal/logger';

import { setStartDate, setEndDate } from 'app/actions';
import { fetchDiff } from 'app/actions/diff';
import { ICommit } from 'app/interfaces';

export function setDates(
  dispatch: (args: any) => void,
  commits: ICommit[],
  selectedPath: string,
  startDate: number,
  endDate: number | Date | null
) {
  const epochStartDate = Math.floor(startDate / 1000);
  const epochEndDate = endDate && Math.floor((endDate as number) / 1000);

  debug('Timeplot: setDates', startDate, epochStartDate, endDate, epochEndDate);
  const [newStartDate, newEndDate] =
    epochStartDate === epochEndDate
      ? [epochStartDate, null]
      : !epochEndDate || epochStartDate < epochEndDate
        ? [epochStartDate, epochEndDate]
        : [epochEndDate, epochStartDate];

  dispatch(setStartDate(newStartDate));
  dispatch(setEndDate(newEndDate));
  dispatch(fetchDiff(selectedPath, commits, newStartDate, newEndDate));
}
