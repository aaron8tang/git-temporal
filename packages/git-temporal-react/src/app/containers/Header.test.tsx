import React from 'react';
import { shallow } from 'enzyme';
import { selectPath } from 'app/actions';

import { Header } from './Header';

const testAuthors = ['Author One', 'Author Two', 'Author Three'];

describe('containers/Header', () => {
  describe('when rendered without props', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      // @ts-ignore
      selectPath.mockClear();
      mockDispatch = jest.fn();
      wrapper = shallow(<Header dispatch={mockDispatch} selectedPath="" />);
    });

    test('it should match snapshot (it should be showing App Name and repository:/)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('when rendered with a proper path', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      // @ts-ignore
      selectPath.mockClear();
      mockDispatch = jest.fn();
      wrapper = shallow(
        <Header dispatch={mockDispatch} selectedPath="some/path/to/File.txt" />
      );
    });

    test('it should match snapshot (it should have path links for all but last part of path)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('when rendered with filteredAuthors', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      // @ts-ignore
      selectPath.mockClear();
      mockDispatch = jest.fn();
      wrapper = shallow(
        <Header dispatch={mockDispatch} filteredAuthors={testAuthors} />
      );
    });

    test('it should match snapshot (it should have path links for all but last part of path)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
