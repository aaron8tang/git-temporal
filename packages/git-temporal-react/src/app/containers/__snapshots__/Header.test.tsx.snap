// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Header when rendered without props it should match snapshot (it should be showing empty indicator) 1`] = `
<div
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 3,
      "display": "flex",
      "marginRight": 10,
      "padding": 10,
    }
  }
>
  <span
    style={
      Object {
        "color": "#333333",
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 20,
        "marginRight": 20,
      }
    }
  >
    Git Temporal 
  </span>
  <span
    style={
      Object {
        "color": "#333333",
        "fontSize": 18,
        "fontWeight": "bold",
        "marginBottom": 20,
        "marginRight": 20,
      }
    }
  >
    Stats for 
  </span>
</div>
`;
